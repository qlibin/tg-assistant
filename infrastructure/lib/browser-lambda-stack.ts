import { Duration, Stack, StackProps, CfnOutput } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as ecr from 'aws-cdk-lib/aws-ecr';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as lambda from 'aws-cdk-lib/aws-lambda';

export interface BrowserLambdaStackProps extends StackProps {
  environmentName: string;
  ecrRepoName: string;
  lambdaName: string;
  imageTag: string; // default provided by app
  tags?: Record<string, string>;
}

export class BrowserLambdaStack extends Stack {
  constructor(scope: Construct, id: string, props: BrowserLambdaStackProps) {
    super(scope, id, props);

    const { environmentName, ecrRepoName, lambdaName, imageTag } = props;

    // Lookup existing ECR repository by name (account/region from env)
    const repository = ecr.Repository.fromRepositoryName(this, 'Repository', ecrRepoName);

    // Execution role for Lambda
    const execRole = new iam.Role(this, 'LambdaExecutionRole', {
      roleName: `browser-lambda-role-${environmentName}`,
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      description: 'Execution role for browser-lambda to pull from ECR and write logs',
    });

    execRole.addManagedPolicy(
      iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole')
    );

    // Allow pull from ECR with least privilege
    execRole.addToPolicy(
      new iam.PolicyStatement({
        actions: ['ecr:GetAuthorizationToken'],
        resources: ['*'],
      })
    );
    execRole.addToPolicy(
      new iam.PolicyStatement({
        actions: [
          'ecr:BatchCheckLayerAvailability',
          'ecr:GetDownloadUrlForLayer',
          'ecr:BatchGetImage',
        ],
        resources: [repository.repositoryArn],
      })
    );

    // Image URI with tag
    const imageUri = `${Stack.of(this).account}.dkr.ecr.${Stack.of(this).region}.amazonaws.com/${ecrRepoName}:${imageTag}`;

    const fn = new lambda.DockerImageFunction(this, 'Function', {
      functionName: lambdaName,
      architecture: lambda.Architecture.ARM_64,
      memorySize: 1024,
      timeout: Duration.seconds(300),
      role: execRole,
      code: lambda.DockerImageCode.fromEcr(repository, {
        tagOrDigest: imageTag,
      }),
      environment: {
        NODE_ENV: 'production',
      },
      logRetention: logs.RetentionDays.ONE_MONTH,
    });

    new CfnOutput(this, 'FunctionName', { value: fn.functionName });
    new CfnOutput(this, 'FunctionArn', { value: fn.functionArn });
    new CfnOutput(this, 'ImageUri', { value: imageUri });
  }
}
