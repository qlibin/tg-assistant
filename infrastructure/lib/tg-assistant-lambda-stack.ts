import { Annotations, CfnOutput, Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as secretsmanager from 'aws-cdk-lib/aws-secretsmanager';

export interface TgAssistantLambdaStackProps extends StackProps {
  environmentName: string;
  lambdaName: string;
  tags?: Record<string, string>;
}

export class TgAssistantLambdaStack extends Stack {
  constructor(scope: Construct, id: string, props: TgAssistantLambdaStackProps) {
    super(scope, id, props);

    const { environmentName, lambdaName } = props;

    // Execution role for Lambda (least privilege: basic execution role)
    const execRole = new iam.Role(this, 'LambdaExecutionRole', {
      roleName: `telegram-webhook-lambda-role-${environmentName}`,
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      description: 'Execution role for Lambda to write logs',
    });

    execRole.addManagedPolicy(
      iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole')
    );

    // Choose code source: use pre-built ZIP when provided by CI, otherwise inline minimal handler to allow synth/tests without filesystem assets
    const lambdaZipPath = process.env.LAMBDA_ZIP_PATH;
    const useInline = !lambdaZipPath;

    const code = useInline
      ? lambda.Code.fromInline(
          'exports.handler = async () => { return { statusCode: 200, body: "ok" }; };'
        )
      : lambda.Code.fromAsset(lambdaZipPath);

    const handler = useInline ? 'index.handler' : 'dist/index.handler';

    // Create or reference the Telegram webhook secret per environment
    const secretName = `/tg-assistant/telegram-webhook-secret/${environmentName}`;
    const telegramWebhookSecret = new secretsmanager.Secret(this, 'TelegramWebhookSecret', {
      secretName,
      description: 'Telegram webhook secret used to validate updates',
      generateSecretString: {
        // Placeholder; value should be updated out-of-band as per runbook
        secretStringTemplate: JSON.stringify({}),
        generateStringKey: 'webhookSecret',
        passwordLength: 32,
        excludePunctuation: true,
      },
    });

    const fn = new lambda.Function(this, 'Function', {
      functionName: lambdaName,
      runtime: lambda.Runtime.NODEJS_22_X,
      architecture: lambda.Architecture.ARM_64,
      memorySize: 1024,
      timeout: Duration.seconds(300),
      role: execRole,
      code,
      handler,
      environment: {
        NODE_ENV: 'production',
        TELEGRAM_WEBHOOK_SECRET_ARN: telegramWebhookSecret.secretArn,
      },
      logRetention: logs.RetentionDays.ONE_MONTH,
    });

    // Grant Lambda permission to read the secret value
    telegramWebhookSecret.grantRead(fn);

    // Context parameter for external API Gateway invoke permission (full SourceArn provided)
    const sourceArnRaw = this.node.tryGetContext('apiGatewaySourceArn') as unknown;
    const sourceArn =
      typeof sourceArnRaw === 'string' && sourceArnRaw.trim().length > 0
        ? sourceArnRaw.trim()
        : undefined;

    if (sourceArn) {
      new lambda.CfnPermission(this, 'ApiGatewayInvokePermission', {
        action: 'lambda:InvokeFunction',
        functionName: fn.functionArn,
        principal: 'apigateway.amazonaws.com',
        sourceArn,
      });
    } else {
      Annotations.of(this).addWarning(
        'API Gateway SourceArn not provided. Skipping Lambda invoke permission. Provide context: -c apiGatewaySourceArn=arn:aws:execute-api:...'
      );
    }

    new CfnOutput(this, 'FunctionName', { value: fn.functionName });
    new CfnOutput(this, 'FunctionArn', { value: fn.functionArn });
    new CfnOutput(this, 'LambdaRegion', { value: Stack.of(this).region });
    if (sourceArn) {
      new CfnOutput(this, 'ApiGatewaySourceArn', { value: sourceArn });
    }
    new CfnOutput(this, 'TelegramWebhookSecretArn', { value: telegramWebhookSecret.secretArn });
  }
}
