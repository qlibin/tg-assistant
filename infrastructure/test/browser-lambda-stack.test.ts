import * as cdk from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import { BrowserLambdaStack } from '../lib/browser-lambda-stack';

// AAA pattern tests for CDK stack

describe('BrowserLambdaStack', () => {
  const baseEnv = { account: '123456789012', region: 'us-east-1' } as const;

  const makeStack = (
    overrides?: Partial<{
      envName: string;
      repo: string;
      lambdaName: string;
      imageTag: string;
    }>
  ) => {
    const app = new cdk.App();

    return new BrowserLambdaStack(app, 'TestStack', {
      env: baseEnv,
      description: 'Test stack',
      environmentName: overrides?.envName ?? 'dev',
      ecrRepoName: overrides?.repo ?? 'qlibin/browser-lambda',
      lambdaName: overrides?.lambdaName ?? 'browser-lambda-dev',
      imageTag: overrides?.imageTag ?? 'sha-abc',
      tags: { app: 'browser-lambda', env: overrides?.envName ?? 'dev' },
    });
  };

  test('synthesizes expected CloudFormation template (snapshot)', () => {
    // Arrange
    const stack = makeStack({ envName: 'dev', imageTag: 'sha-abc' });

    // Act
    const templateJson = Template.fromStack(stack).toJSON();

    // Assert
    expect(templateJson).toMatchSnapshot();
  });

  test('creates DockerImage Lambda with expected config', () => {
    // Arrange
    const stack = makeStack({ envName: 'dev', imageTag: 'latest' });

    // Act
    const template = Template.fromStack(stack);

    // Assert
    template.hasResourceProperties('AWS::Lambda::Function', {
      FunctionName: 'browser-lambda-dev',
      MemorySize: 1024,
      Timeout: 300,
      Architectures: ['arm64'],
      Environment: {
        Variables: {
          NODE_ENV: 'production',
        },
      },
    });
  });

  test('grants least-privilege ECR pull permissions to execution role', () => {
    // Arrange
    const stack = makeStack();
    const template = Template.fromStack(stack);

    // Assert
    template.hasResourceProperties('AWS::IAM::Policy', {
      PolicyDocument: {
        Statement: Match.arrayWith([
          Match.objectLike({
            Action: 'ecr:GetAuthorizationToken',
            Resource: '*',
          }),
          Match.objectLike({
            Action: Match.arrayWith([
              'ecr:BatchCheckLayerAvailability',
              'ecr:GetDownloadUrlForLayer',
              'ecr:BatchGetImage',
            ]),
          }),
        ]),
      },
    });
  });

  test('configures log retention for 30 days', () => {
    // Arrange
    const stack = makeStack();
    const template = Template.fromStack(stack);

    // Assert
    template.hasResourceProperties('Custom::LogRetention', {
      RetentionInDays: 30,
    });
  });
});
