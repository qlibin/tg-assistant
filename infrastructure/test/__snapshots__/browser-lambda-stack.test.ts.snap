// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrowserLambdaStack synthesizes expected CloudFormation template (snapshot) 1`] = `
{
  "Description": "Test stack",
  "Outputs": {
    "FunctionArn": {
      "Value": {
        "Fn::GetAtt": [
          "Function76856677",
          "Arn",
        ],
      },
    },
    "FunctionName": {
      "Value": {
        "Ref": "Function76856677",
      },
    },
    "ImageUri": {
      "Value": "123456789012.dkr.ecr.us-east-1.amazonaws.com/qlibin/browser-lambda:sha-abc",
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Function76856677": {
      "DependsOn": [
        "LambdaExecutionRoleDefaultPolicy6D69732F",
        "LambdaExecutionRoleD5C26073",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "ImageUri": {
            "Fn::Join": [
              "",
              [
                "123456789012.dkr.ecr.us-east-1.",
                {
                  "Ref": "AWS::URLSuffix",
                },
                "/qlibin/browser-lambda:sha-abc",
              ],
            ],
          },
        },
        "Environment": {
          "Variables": {
            "NODE_ENV": "production",
          },
        },
        "FunctionName": "browser-lambda-dev",
        "MemorySize": 1024,
        "PackageType": "Image",
        "Role": {
          "Fn::GetAtt": [
            "LambdaExecutionRoleD5C26073",
            "Arn",
          ],
        },
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "FunctionLogRetention5FDF6B4D": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "Function76856677",
              },
            ],
          ],
        },
        "RetentionInDays": 30,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "LambdaExecutionRoleD5C26073": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Execution role for browser-lambda to pull from ECR and write logs",
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "RoleName": "browser-lambda-role-dev",
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaExecutionRoleDefaultPolicy6D69732F": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "ecr:GetAuthorizationToken",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":ecr:us-east-1:123456789012:repository/qlibin/browser-lambda",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaExecutionRoleDefaultPolicy6D69732F",
        "Roles": [
          {
            "Ref": "LambdaExecutionRoleD5C26073",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "2819175352ad1ce0dae768e83fc328fb70fb5f10b4a8ff0ccbcb791f02b0716d.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
